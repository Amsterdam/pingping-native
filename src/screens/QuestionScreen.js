import React, {useEffect, useRef, useState} from 'react';
import {StyleSheet} from 'react-native';
import {useMutation, useQuery} from '@apollo/client';
import PropTypes from 'prop-types';
import {View as AnimatableView} from 'react-native-animatable';
import ErrorComponent from '../components/shared/ErrorComponent';
import QuestionSkeleton from '../components/skeleton/QuestionSkeleton';
import UPDATE_TASK_MUTATION from '../apollo/Mutation/updateTaskMutation';
import REVERT_TASK_MUTATION from '../apollo/Mutation/revertTaskMutation';
import GET_STATUS_QUERY from '../apollo/Query/getStatusQuery';
import QuestionComponent from '../components/onboarding/QuestionComponent';
import {
  revertTaskFunc,
  setRevertedQuestionValues,
  submitAnswer,
  updateConfirmTask,
} from '../helpers/questionAnswerHelpers';

const INITIAL_STATE = {
  answerSelected: false,
  day: '',
  month: '',
  year: '',
  choices: [],
};

const QuestionScreen = ({navigation}) => {
  const {data, loading, error, refetch} = useQuery(GET_STATUS_QUERY);
  const [updateTask] = useMutation(UPDATE_TASK_MUTATION);
  const [revertTask] = useMutation(REVERT_TASK_MUTATION);
  const [loadingQuestion, setLoadingQuestion] = useState(false);
  const animationRef = useRef(null);
  const current = data?.getStatus?.currentTask;
  const answeredBefore = current?.answer;
  const currentTask = current?.task;
  const previousTask = data?.getStatus?.previousTask?.task;
  const [state, setState] = useState(INITIAL_STATE);

  useEffect(() => {
    if (answeredBefore) {
      setRevertedQuestionValues(currentTask, answeredBefore, setState);
    }
  }, [answeredBefore, currentTask, navigation]);

  if (error) {
    return (
      <ErrorComponent
        functionToRetry={refetch}
        somethingWentWrong
        onPress={() => navigation.goBack()}
      />
    );
  }

  if (data && currentTask) {
    const doRevertTask = () => {
      revertTaskFunc(
        setLoadingQuestion,
        previousTask,
        navigation,
        refetch,
        revertTask,
        animationRef,
      );
    };

    const doUpdateTask = () => {
      submitAnswer(
        currentTask,
        state,
        setLoadingQuestion,
        updateTask,
        setState,
        refetch,
        INITIAL_STATE,
        animationRef,
      );
    };

    const doUpdateConfirmTask = (answer) => {
      updateConfirmTask(
        answer,
        updateTask,
        currentTask,
        refetch,
        animationRef,
        setLoadingQuestion,
      );
    };

    return (
      <AnimatableView
        style={styles.flex}
        duration={400}
        ref={animationRef}
        useNativeDriver>
        {!loadingQuestion && (
          <QuestionComponent
            currentTask={currentTask}
            updateTask={updateTask}
            refetch={refetch}
            doRevertTask={doRevertTask}
            state={state}
            setState={setState}
            doUpdateTask={doUpdateTask}
            setLoadingQuestion={setLoadingQuestion}
            doUpdateConfirmTask={doUpdateConfirmTask}
          />
        )}

        {(loadingQuestion || loading) && <QuestionSkeleton />}
      </AnimatableView>
    );
  }
  return <QuestionSkeleton />;
};

const styles = StyleSheet.create({
  flex: {flex: 1},
});

QuestionScreen.propTypes = {
  navigation: PropTypes.object.isRequired,
};

export default QuestionScreen;
