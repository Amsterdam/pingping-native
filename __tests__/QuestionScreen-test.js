import React from 'react';
import {MockedProvider} from '@apollo/client/testing';
import QuestionScreen from '../src/screens/QuestionScreen';

import renderer from 'react-test-renderer';

test('renders correctly', () => {
  const tree = renderer
    .create(
      <MockedProvider mocks={[]}>
        <QuestionScreen />
      </MockedProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
