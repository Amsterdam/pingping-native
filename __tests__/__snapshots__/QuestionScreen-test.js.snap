// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "paddingHorizontal": 20,
      "paddingVertical": 50,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 35,
        "marginTop": 25,
      }
    }
  >
    <View
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 10,
              "height": 30,
              "marginTop": 10,
              "width": 30,
            },
            Object {
              "backgroundColor": "#E1E9EE",
              "overflow": "hidden",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "translateX": -350,
                },
              ],
            }
          }
        >
          <BVLinearGradient
            colors={
              Array [
                4292995566,
                4294113532,
                4292995566,
              ]
            }
            endPoint={
              Object {
                "x": 1,
                "y": 0,
              }
            }
            locations={null}
            startPoint={
              Object {
                "x": 0,
                "y": 0,
              }
            }
            style={
              Object {
                "flex": 1,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 10,
              "height": 20,
              "marginTop": 10,
              "width": 100,
            },
            Object {
              "backgroundColor": "#E1E9EE",
              "overflow": "hidden",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "translateX": -350,
                },
              ],
            }
          }
        >
          <BVLinearGradient
            colors={
              Array [
                4292995566,
                4294113532,
                4292995566,
              ]
            }
            endPoint={
              Object {
                "x": 1,
                "y": 0,
              }
            }
            locations={null}
            startPoint={
              Object {
                "x": 0,
                "y": 0,
              }
            }
            style={
              Object {
                "flex": 1,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 10,
              "height": 20,
              "marginTop": 10,
              "width": 70,
            },
            Object {
              "backgroundColor": "#E1E9EE",
              "overflow": "hidden",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "translateX": -350,
                },
              ],
            }
          }
        >
          <BVLinearGradient
            colors={
              Array [
                4292995566,
                4294113532,
                4292995566,
              ]
            }
            endPoint={
              Object {
                "x": 1,
                "y": 0,
              }
            }
            locations={null}
            startPoint={
              Object {
                "x": 0,
                "y": 0,
              }
            }
            style={
              Object {
                "flex": 1,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "space-between",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "borderRadius": 10,
                "height": 40,
                "marginTop": 10,
                "width": "100%",
              },
              Object {
                "backgroundColor": "#E1E9EE",
                "overflow": "hidden",
              },
            ]
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateX": -350,
                  },
                ],
              }
            }
          >
            <BVLinearGradient
              colors={
                Array [
                  4292995566,
                  4294113532,
                  4292995566,
                ]
              }
              endPoint={
                Object {
                  "x": 1,
                  "y": 0,
                }
              }
              locations={null}
              startPoint={
                Object {
                  "x": 0,
                  "y": 0,
                }
              }
              style={
                Object {
                  "flex": 1,
                }
              }
            />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "borderRadius": 10,
                "height": 40,
                "marginTop": 6,
                "width": "80%",
              },
              Object {
                "backgroundColor": "#E1E9EE",
                "overflow": "hidden",
              },
            ]
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateX": -350,
                  },
                ],
              }
            }
          >
            <BVLinearGradient
              colors={
                Array [
                  4292995566,
                  4294113532,
                  4292995566,
                ]
              }
              endPoint={
                Object {
                  "x": 1,
                  "y": 0,
                }
              }
              locations={null}
              startPoint={
                Object {
                  "x": 0,
                  "y": 0,
                }
              }
              style={
                Object {
                  "flex": 1,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View>
      <View
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "borderRadius": 5,
                "height": 50,
                "marginTop": 6,
                "width": "100%",
              },
              Object {
                "backgroundColor": "#E1E9EE",
                "overflow": "hidden",
              },
            ]
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateX": -350,
                  },
                ],
              }
            }
          >
            <BVLinearGradient
              colors={
                Array [
                  4292995566,
                  4294113532,
                  4292995566,
                ]
              }
              endPoint={
                Object {
                  "x": 1,
                  "y": 0,
                }
              }
              locations={null}
              startPoint={
                Object {
                  "x": 0,
                  "y": 0,
                }
              }
              style={
                Object {
                  "flex": 1,
                }
              }
            />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "borderRadius": 5,
                "height": 50,
                "marginTop": 6,
                "width": "100%",
              },
              Object {
                "backgroundColor": "#E1E9EE",
                "overflow": "hidden",
              },
            ]
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateX": -350,
                  },
                ],
              }
            }
          >
            <BVLinearGradient
              colors={
                Array [
                  4292995566,
                  4294113532,
                  4292995566,
                ]
              }
              endPoint={
                Object {
                  "x": 1,
                  "y": 0,
                }
              }
              locations={null}
              startPoint={
                Object {
                  "x": 0,
                  "y": 0,
                }
              }
              style={
                Object {
                  "flex": 1,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View>
      <View
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignSelf": "flex-end",
                "borderRadius": 6,
                "height": 25,
                "marginTop": 6,
                "width": 100,
              },
              Object {
                "backgroundColor": "#E1E9EE",
                "overflow": "hidden",
              },
            ]
          }
        >
          <View
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateX": -350,
                  },
                ],
              }
            }
          >
            <BVLinearGradient
              colors={
                Array [
                  4292995566,
                  4294113532,
                  4292995566,
                ]
              }
              endPoint={
                Object {
                  "x": 1,
                  "y": 0,
                }
              }
              locations={null}
              startPoint={
                Object {
                  "x": 0,
                  "y": 0,
                }
              }
              style={
                Object {
                  "flex": 1,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
